packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -6278,50 +6281,197 @@ abstract class RenderObjectElement extends Element {
-            ErrorDescription('However, a RenderObject can only receive parent data from at most one ParentDataWidget.'),
-            ErrorHint('Usually, this indicates that at least one of the offending ParentDataWidgets listed above is not placed directly inside a compatible ancestor widget.'),
-            ErrorDescription('The ownership chain for the RenderObject that received the parent data was:\n  ${debugGetCreatorChain(10)}'),
+            ErrorDescription(
+              'Multiple ParentDataWidgets are providing competing ParentData '
+              'to the same RenderObject:'
+            ),
+            for (final ParentDataElement<ParentData> ancestor in getCompetingParentData())
+              ErrorDescription(
+                '- ${ancestor.widget} which writes ParentData of type '
+                '${ancestor.parentDataType}'
+              ),
+            ErrorDescription(
+              'A RenderObject can receive parent data from multiple '
+              'ParentDataWidgets, but the Type of ParentData must be unique to '
+              'prevent one overwriting another.'
+            ),
+            ErrorHint(
+              'This can be resolved by using mixins to separate aspects of a '
+              'ParentData class into separate types, mixing them all in on the '
+              'full ParentData Object, such as KeepAlive does with '
+              'KeepAliveParentDataMixin.'
+            ),
+            ErrorDescription(
+              'The ownership chain for the RenderObject that received the '
+              'parent data was:\n  ${debugGetCreatorChain(10)}'
+            ),
           ]);
-        } on FlutterError catch (e) {
-          _reportException(ErrorSummary('while looking for parent data.'), e, e.stackTrace);
+        } on FlutterError catch (error) {
+          _reportException(
+            ErrorSummary('while looking for parent data.'),
+            error,
+            error.stackTrace,
+          );
         }
       }
       return true;
     }());
+  }
+
+  List<ParentDataElement<ParentData>> _findAncestorParentDataElements() {
+    Element? ancestor = _parent;
+    final List<ParentDataElement<ParentData>> result = <ParentDataElement<ParentData>>[];
+    final Set<Type> debugAncestorTypes = <Type>{};
+    final Set<Type> debugParentDataTypes = <Type>{};
+    final List<Type> debugAncestorCulprits = <Type>[];
+    final List<Type> debugParentDataCulprits = <Type>[];
+
+    // More than one ParentDataWidget can contribute ParentData, but there are
+    // some constraints.
+    // 1. ParentData can only be written by unique ParentDataWidget types.
+    //    For example, two KeepAlive ParentDataWidgets trying to write to the
+    //    same child is not allowed.
+    // 2. Each contributing ParentDataWidget must contribute to a unique
```

Comment:
Do we actually need to specifically check for case 1? Seems like case 1 is already covered if we enforce case 2? Could we just check for case 2 to simplify the logic?

Reply:
Yeah, I think I misunderstood your comment in as separate case in https://github.com/flutter/flutter/pull/133581#discussion_r1332029125 
I can simplify a bunch then.